datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1335,39 +1335,16 @@ private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
       // and schedule a retry. The retry should be able to diff the remaining ZooKeeper work
       if (_config.getEnableAssignmentTokens()) {
         _adapter.updateAllAssignmentsAndIssueTokens(newAssignmentsByInstance, stoppingDatastreamGroups);
+        try {
+          _tokenClaimExecutor.submit(() -> waitForStopToPropagateAndMarkDatastreamsStopped(stoppingDatastreamGroups));
+        } catch (RejectedExecutionException ex) {
+          _log.error("handleLeaderDoAssignment: Failed to submit the task that waits for stop propagation");
+          succeeded = false;
+        }
+
       } else {
         _adapter.updateAllAssignments(newAssignmentsByInstance);
-      }
-
-      Set<String> failedStreams = Collections.emptySet();
-      // Poll the zookeeper to ensure that hosts claimed assignment tokens for stopping streams
-      if (_config.getEnableAssignmentTokens() &&
-          !PollUtils.poll(() -> _adapter.getNumUnclaimedTokensForDatastreams(stoppingDatastreamGroups) == 0,
-            STOP_PROPAGATION_RETRY_MS, _config.getStopPropagationTimeout())) {
-        Map<String, List<AssignmentToken>> unclaimedTokens =
-            _adapter.getUnclaimedAssignmentTokensForDatastreams(stoppingDatastreamGroups);
-        failedStreams = unclaimedTokens.keySet();
-        Set<String> hosts = unclaimedTokens.values().stream().flatMap(List::stream).
-            map(AssignmentToken::getIssuedFor).collect(Collectors.toSet());
-        _log.error("Stop failed to propagate within {}ms for streams: {}. The following hosts failed to claim their token(s): {}",
-            _config.getStopPropagationTimeout(), failedStreams, hosts);
-        revokeUnclaimedAssignmentTokens(unclaimedTokens, stoppingDatastreamGroups);
-        _metrics.updateMeter(CoordinatorMetrics.Meter.NUM_FAILED_STOPS, failedStreams.size());
-      }
-
-      // TODO Explore introduction of UNKNOWN/WARN state for streams that failed to stop
-
-      for (DatastreamGroup datastreamGroup : stoppingDatastreamGroups) {
-        for (Datastream datastream : datastreamGroup.getDatastreams()) {
-          // Only streams that were confirmed to have stopped successfully will be transitioned to STOPPED state
-          if (!failedStreams.contains(datastream.getName())) {
-            datastream.setStatus(DatastreamStatus.STOPPED);
-            if (!_adapter.updateDatastream(datastream)) {
-              _log.warn("Failed to update datastream: {} to stopped state", datastream.getName());
-              succeeded = false;
```

Comment:
My main concern with this pull request is that earlier failures to update Zookeeper state would cause the assignment change to be rescheduled.

Now that I moved this code to a separate thread, I can't propagate the result back to handleLeaderDoAssignment to schedule a retry. I'm not sure how valuable this check is. It seems that if Zookeeper failed during a stream update, a retry won't do much help here.

If you think that this retry path is indeed necessary, I can find a workaround to solve this. One way is to introduce a new type of event for STOPPING -> STOPPED state transition. I can schedule that in the main event queue and have it rescheduled until it succeeds.

Share your thoughts @vmaheshw @shrinandthakkar

Reply:
The retry path is necessary. I like the workaround approach. @shrinandthakkar is working on some changes that will be conflicting with this path.

Reply:
After assessing the changes to add a new event type for STOPPING -> STOPPED transition, I realized that it is a large change and has risks of introducing regressions.
Instead of doing that I added a retry in the new separate thread that confirms the streams have stopped and updates the Zookeeper. This change won't affect the old behavior/code since it will only trigger when the assignment tokens feature is enabled. This will unblock me and let me test the feature.

Reply:
Discussed offline with @vmaheshw. Decided to go forward with retry poll for `STOPPED` → `STOPPING` transition instead of a separate event type. Added improvements.